### PR TITLE
SPARK-2533 - Add locality levels (with tasks count) on job stages summary

### PR DESCRIPTION
**Pull request by jbonofre**

### CODE REVIEW

**Comment by rxin** [comment]
Can you post a screenshot for the change?

**Comment by jbonofre** [comment]
Sure, give me 2 minutes ;)

**Comment by jbonofre** [comment]
screen1
screen2

**Comment by jbonofre** [comment]
Note the "new" Locality Levels column in the stage summary view. The format is:

locality_level(number_of_tasks)

For instance NODE_LOCAL(1), means one task in this stage with NODE_LOCAL locality level.

**Comment by rxin** [comment]
cc @pwendell @kayousterhout  thoughts on this?

**Comment by pwendell** [comment]
ping @kayousterhout to me it's the normal issue - sure it's useful for some cases, but is it worth putting in the index page? It could be better to just put a locality summary on the stage page itself, if one doesn't already exist.

**Comment by jbonofre** [comment]
@pwendell the locality summary is on the stages page. I can also do a locality summary on the view of each stage (we have it in the tasks table, but a summary with the tasks count for each locality is helpful IMHO).

**Comment by pwendell** [comment]
Yeah this was my thought - could we provide a summary in the stage page rather than in the stage index.    I do see how an aggregated summary is significantly more useful than just mentally aggregating based on the task table. Doing this on the stage page would avoid adding more columns to the index view, and this column could get complicated if stages have many locality levels in play. It might be nice to see an alternative patch that does that.

**Comment by jbonofre** [comment]
+1, let me update this PR according to @pwendell comment. Give me 10mn ;) Thanks Patrick !

**Comment by jbonofre** [comment]
screen

**Comment by jbonofre** [comment]
Updated PR by moving locality level summary on stage details.

**Comment by jbonofre** [comment]
Gently reminder ... Thanks !

**Comment by kayousterhout** [comment]
Two suggestions:

1) can we change the format of this to say "Process local: 2 tasks; Rack local: 4 tasks; Any: 15 tasks"? I think that's easier to read and more clear what it is.

2) I wonder if we should put this under the "Show additional metrics" drop down. Right now, it's one of the first things someone will see when they open the UI, and it's not clear to me that that's the right prioritization.

File: core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala [hunk 1]
```diff
@@ -21,6 +21,7 @@ import java.net.URLEncoder
 import java.util.Date
 import javax.servlet.http.HttpServletRequest
 
+import scala.collection.mutable
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: Can you directly import mutable.HashMap here? That's more consistent with style elsewhere

File: core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala [hunk 2]
```diff
@@ -122,6 +123,24 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
       val externalAccumulables = allAccumulables.values.filter { acc => !acc.internal }
       val hasAccumulators = externalAccumulables.size > 0
 
+      val tasksLocality = new mutable.HashMap[String, Long]
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: Can you move all of this into a "getLocalitySummaryString" function or similar?

File: core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala [hunk 3]
```diff
@@ -122,6 +123,24 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
...
+      if (stageData.taskData.size > 0) {
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: Is this outer if-statement necessary?

File: core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala [hunk 4]
```diff
@@ -122,6 +123,24 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
...
+        for ((key, value) <- stageData.taskData) {
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: Can you change this to stageData.taskData.values.foreach { taskUiData => .... }

File: core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala (hunk 2 above)
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: rename to localityCounts or localityLevelToNumTasks?

File: core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala [hunk 5]
```diff
@@ -122,6 +123,24 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
...
+          val locality = value.taskInfo.taskLocality.toString;
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: you can re-write the next few lines as:
tasksLocality.put(locality, tasksLocality.getOrElse(locality, 0) + 1)

File: core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala [hunk 6]
```diff
@@ -122,6 +123,24 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
...
+          if (tasksLocality.contains(locality)) {
+            val count: Long = tasksLocality.get(locality).get + 1;
+            tasksLocality.put(locality, count);
+          } else {
+            tasksLocality.put(locality, 1);
+          }
+        }
+      }
+      var localityLevels: String = "";
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: How about re-writing this as: (all of the concat statements are tough to read)

tasksLocality.map { _ match {
  case (localityLevel, count) =>
    s"$localityLevel ($count)"
}}.mkString("; ")

File: core/src/main/scala/org/apache/spark/ui/jobs/StageTable.scala [hunk 7]
```diff
@@ -19,6 +19,7 @@ package org.apache.spark.ui.jobs
 
 import java.util.Date
 
+import scala.collection.mutable.{HashMap}
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/ui/jobs/StageTable.scala]: remove this?

File: core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala (hunk 5 above)
Comment (jbonofre) [inline core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: @kayousterhout  Thanks so much for all your feedbacks !!! Much appreciated. Let me update the PR accordingly.

**Comment by jbonofre** [comment]
I rebased and updated the PR to improve the code according to Kay's comments.

Now, I'm starting to implement the two Kay's suggestions.

@kayousterhout thanks again for your support !

**Comment by jbonofre** [comment]
Rebase

**Comment by jbonofre** [comment]
Sorry for the noise

**Comment by jbonofre** [comment]
Crap I did a mistake, let me clean it up.

**Comment by jbonofre** [comment]
Let me close this PR and create a new one (easier to read).

**Comment by kayousterhout** [comment]
Would you mind re-opening this and closing the new one?  You can change the code here to be equivalent to the code in your new branch by doing "git push --force origin SPARK-2533-2:SPARK-2533".  It's easier for us to track things if we keep the history in one pull request.

As an aside, when you're working on pull requests, it's helpful if you implement changes as new commits on top of your original one.  That way, reviewers can review the commits in the history to look at just what's changed since the last review.  When we merge things into Spark, we squash it all into one commit, so using a bunch of commits in a pull request won't pollute the git history later on.

**Comment by jbonofre** [comment]
OK no problem. I thought you would prefer a new clean one. My bad and sorry about that.

**Comment by jbonofre** [comment]
Unfortunately, github says the branch has been recreated or forced-pushed :/

**Comment by jbonofre** [comment]
And yes, I did a mistake (merge instead of rebase).

**Comment by jbonofre** [comment]
@kayousterhout I'm very sorry about my mistake (I worked on another project in the same time and mixed the git actions :(, probably too early on the morning for me ;)).

**Comment by kayousterhout** [comment]
No problem -- at this point, just do whatever's easiest to get a pull request that's in an ok state!

**Comment by jbonofre** [comment]
Thanks @kayousterhout. I promise it's my last mistake ;)

**Pull request closed without merging**